studio/lib/fetchWithToken.ts

```diff
@@ -0,0 +1,38 @@
+/**
+ * fetchWithToken
+ *
+ * Utility function to securely fetch data from the server using an API route.
+ * This function sends a POST request to the /api/fetchData route with a GROQ query
+ * and optional parameters. It returns the fetched data as a Promise of type T.
+ *
+ * This function should be used in client-side components where data needs to be
+ * fetched securely using the server-side token without exposing it in the client code.
+ *
+ * @template T - The expected return type of the fetched data.
+ * @param {string} query - The GROQ query string to be executed on the server.
+ * @param {Record<string, any>} [params] - Optional parameters to be passed along with the query.
+ * @returns {Promise<T>} - A promise that resolves to the fetched data of type T.
+ *
+ * @throws Will throw an error if the fetch operation fails or if the response is not ok.
+ *
+ * @example
+ * const data = await fetchWithToken<MyDataType>("*[_type == 'myType']", { param1: 'value' });
+ */
+export const fetchWithToken = async <T>(
+  query: string,
+  params?: Record<string, any>
+): Promise<T> => {
+  const response = await fetch("/api/fetchData", {
```

Comment:
Could we use [Server Actions](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations) here?

Reply:
hmm.. could be that I gave it a try only with client.fetch and not loadQuery! I'll give it a test run, cause it'd be nice to omit this as well! 🎉

Reply:
ok so I looked into it, the reason why it's not working is because this is a custom React component within Sanity. and server actions, such as the "use server" directive, are specifically designed for use within Next.js server components, not client-side React components like the one we’re using in Sanity 💡

Reply:
Makes sense if Sanity Studio is not using Next.js, so api route is the way to go here 👍 (but server actions can definitely be used in Next.js client components, so could be useful later)